Custom Anchor - Dynamic Group "Group by Frame" - Border Settings - Groups thumbnails

mrbuds opened the pull request:
# Description
- Killed group & dynamicgroup thumbnail, use the animated icons instead, with ability to set your own
![img](https://i.imgur.com/1WsEQXz.gif)
![img](https://i.imgur.com/ldgbIwq.png)
- Renamed and reorganized a few options in Group and Display tab
https://i.imgur.com/hJIWwj3.png
https://i.imgur.com/sfXrmLt.png
- Use same border settings for both groups and dynamic groups and changed default settings for something closer to pixel perfect style
- New dependency LibGetFrame, publicly accessible with WeakAuras.GetUnitFrame documentation at <https://github.com/mrbuds/LibGetFrame/blob/master/README.md>
- All regions can use a custom anchor function, custom function needs to return a frame
```Lua
function()
  return SUFUnitplayer
end
```
![img](https://i.imgur.com/PDaSIgY.png)
- Dynamic group's growers have a new option "Group by Frame", it can be "unit frame", "nameplates" or a custom function. Enabling this option disable border options.
![img](https://i.imgur.com/Ls0mW5G.png)
Custom anchor per unit example :
```Lua
function(frames, activeRegions)
    for _, regionData in ipairs(activeRegions) do
        local unit = regionData.region.state and regionData.region.state.destUnit
        if unit then
            local frame = C_NamePlate.GetNamePlateForUnit(unit)
            if frame then
                frames[frame] = frames[frame] or {}
                tinsert(frames[frame], regionData)
            end
        end
    end
end
```
- Custom grower interface is incompatible with previously written custom grow, but easy to adapt
It now needs to return an array keyed by anchor frame, or "" if anchor is dynamic group region, followed by an array keyed by regionData `[anchor][regionData] = { x , y }`

Custom grower example:
```Lua
function(newPositions, activeRegions)
    local frames = {}
    for _, regionData in ipairs(activeRegions) do
        local unit = regionData.region.state and regionData.region.state.unit
        if unit then
            local frame = C_NamePlate.GetNamePlateForUnit(unit)
            if frame then
                frames[frame] = frames[frame] or {}
                tinsert(frames[frame], regionData)
            end
        end
    end
    for frame, regionsData in pairs(frames) do
        local totalWidth = #regionsData - 1
        for _, regionData in ipairs(regionsData) do
            totalWidth = totalWidth + (regionData.data.width or regionData.region.width)
        end
        local x, y = - totalWidth/2, - (#regionsData - 1)/2
        newPositions[frame] = {}
        for i, regionData in ipairs(regionsData) do
            x = x + (regionData.data.width or regionData.region.width) / 2
            newPositions[frame][regionData] = { x, y }
            x = x + (regionData.data.width or regionData.region.width) / 2
        end
    end
end
```

Any suggestion, advice or critic welcome

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update
- [x] This break compatibility with previous custom code

# How Has This Been Tested?
No custom code: show buffs as grid in unitframes
```
!nBvZYTTnq4NMEOhAhrh70Cv)qLOoYuQeuTjxGeijejQHaujaLT8b9S3DbinPuLDDMEPZ4HgC5I93V9BfnGMqjgkzc)aLSJsYX)arfv6hPKphpBcLKQRY5vJyzpKxP3tjJKINFMvLFAtIwlTcq0tl2U1WT0ObuYXENzQSsD1sTqzPKKflNhonP1GJ1sDf4ROau1OBCp)G75TOq4g1g(qVf4vRuclDeyYmRqRmU42YQSWbqtb(v3PTWrtP7m8MTsuuWRmE)K8Ia8yopTE72KJ75uYxcNVC6Q5W3DVYQRy3aoypxkNLB8w2uNYpWvwcClXtuY6XdjjRjjdJbJvJbazVKDem(lQUSI7uLSmCoyCNmWzCM0wsjk2oUPnsRvDHM1lktlR3Pi7zzCm6HK0YRum5VdjeudGAwawIf7yyjPFfPjnuAf8mhsguH1EHgEMwb5e4GDmH69R9lv23N(WnysrHIsghgLeg7IoxcIidOcfZlA6KGQsnZb8Gw(6mjZyOPU6FgkCxnaZAlu(VEPuJ4z(LcbXPnO2(GTahylWb2cCGTb)8DUW1vk5ONz1wa2TyVTlcnzmPVp4XVOA(tH5fqMhCAZsXtCjLu5smpWk)i0LfzW8u9(w9jyScXq74rd4MSkAwY04H3hcMGLlQnGsdGAfml2bcKIDGUr3HZOICVtINGGffuFv2PWacKQK4HtMTIGx26Hhyv37)2b0BBfmt5EhSUvNDOfD9PboVlaobM1YYkpTbr5N2y1(tBRaamGlW)bdwmlJg9rFC9hICaHJEWNJtrD8blzCCyyeIuKBVKzasITIcF9UgD88e(VwfV)h)t5kWuvzZ5Qc0WF4JdCyL5UQrQ7M5I(DlFbdhI1WiHhSIONxyckKafx67K7QfcNbmE56hvKhfqUmQtGddawtabck)TiEmL6hhlHaY408mgN3IJ5c6JxHK5CsRooMlOOUg73vOHC8Th4jEH3RZHI6pfmWZ5Ef6iSQ(1ZQI))LEkNByw87Uohg5VnnrqZwGoIRMwl)j7Atjl)8uVXJ1P43B0YF8hmnVM9V7pNEB1yJFAL45tB(TAwooRbRFp33VJfQ90(B9JuN4U5XHRsw0ZZEqYIvjZNff2i)SPJzrrHXRhTijzX9XZ(8xsANE(hobnNN(l4MgUjEIRMeTaTT)AF9YRDqyePsCGZ14k5IIsKW6UUfh)33vGaGPv8)QMRYoIlfU5URsiI65ljP1wRwTauagQA(qPi7byY24Ot7vtBzq6VCWrw4U2CbYge9PM3Ay5c8KCDiZ0ZPi)((Tqp65KXI2zu2bxDD3ZA9o)eCTvJ8uyuqEzhiso7wQDAdYJy6ycX25eHHb9RC3pytUV07f0c9reVd0VBjq2H73DS42vb)Y3WSd7h(L0V2QLh489draHngh(V2(H2OTnZg1Ut3vQoaU)2BWiG(3
```

custom "anchor per unit": attach icon to nameplate (note: this would be better to achieved without dynamic group using custom anchor from aura region)
```
!nF12UnUnt4NMcKfOnWk7bS30lSDS31TELDnvAtBrLfLeTmrKj9ssLtfvp7)ZmuYw2ojB(rVPWjouJgoh)MdjoiokMzJzxkUnMTjMLJ)aKkm67IzFE2Ij)XSWO(tJzPAtUWmGNDtUrVnMnOu(4JCtEDsKwx6KaP7NTALv4Id7fZEOZ5cJmp6HTIy2IlBf0qDP2aAkmazj8c673sF)oKOlMbMaBlptGpY4zoPwzjZ1Xno4aWYkPsAx7pdh9uHZoJSOqySEfeTJaEmxKwTAL3E(8OPZhFf4Do6rELHFbOGTIYYj5wVCTvPIBfkhdUL8(y2YH9zrlzr9xacRcvkBBj)bq4mfFJO5w0vafi4LU17fYCJGecB(OPtr(Qu7nnN34Z0LvBu7DDdpxwzHyupiwYvYnCmu0nsWYbdhjU07hwrMwb2FJBP0kbk4nCP61Z9Uy7RJF4g8szHkMnCuy0OfK1rogcba)FHOOjdcSwQ5eoRYkwMvYT24ukUNHe3ub4P2uP)Tht1kFuCmrGCAd8Sl6kGqxbe6kGqx9o)9uOxcO9R)57)sGzY1t6TPn4kiBjJx2G8QCR1MzBD7TEpeoEq7Pr5fqGiOozU8ErjKXix1dXYFaqfYmOGQABl)m06bRcuygi75cZveoA4vSOzFHIktLBakOHOnyO1Mju5svb9Y(hCTbyPIZdkWynKDGqVYngQzGOaBr)lNCfJOVsw01dARrFhKqi9HNC6SBHshsAFShb6Waf354zRRtYfwhQ26eNUobr8a6hIzSvg4muOXD84Wp4R6)nzoG(rP690Xip(4cB4IrJcr0u5Q5AjwS0cC8g3efzBygWKnvOkqj92pGoyL1P3CyqGTQsrTioJSd73xNGTmUv0a7Et9pvN4)SsBQtwcm4ttxcgCDIub)ULln2Zo6Ej5693T9tPgWh1jvuG4h7kQZ9hphsbobyeQ8N9TNVlwEI8LRAfUBTqD67pYoiFMmKHldHJZXuY5Fs429WynfNodf6BEE5H6TryVSI3fmPO9Fs)5Vil4ikyW(V)NVTGCsiFBCNDW1pij9cMnuBC6lpGi9a81UAUdHCyhnzNkCRV77f4yfNWO4L)ABnXf0WPSDJwkkHrLPhoX7z0HUmxFNIDNekagSNa1(aKGeScK(ln9YUwF3WsOLRL403qgatWaIMMPy74b4GLMrqq)l9(E4qBQDdNgYTUJhHHYGNcnDs)IoNe0Htd7mIdzTQPb0tpF7LND(udIpAI4HgNFidvDg5zHSXWFiONFbGtZuSBeIT9XblUfO)JH5)lpcf6ia5stZqieA9YJYcA2qz)W1geJ4E3s7AoaVUUdUSr7vP47B4YF87SnpM9T1hX3kncHgBKpwN8lv8Cd1gn6qD)k2YRd3)ExlLiVF2q)RIM1rZEuZSRIMojCud9dk6MegoAXYbZIGbQlM8Pph1wuEIsUvALPLIDJrfruajCgky)DU(47GMGFiEWf(e3AHSyTdxsB)Yn)73NbbaJnIVwjuzpGlUCX7FYb0iF(qsALZPvZagGAMMxSwMDd0WW(CRZGm11f7UhF3vzO(tyz)ECz6)VBMtlDno)YVwT5tfb3ZEYfaUZV2agl5LBxZjdhv6uj24l8JXSh16nnB74BIIzTlLwoKkZXCj12BG)ASDlZHA33nPojJA)9kW7hSztWlSldQRU4RUBCjvysJkRHiy7)ZutrE76nb9oDuuRd26md8lFM6JN3cx6D0iP4)3p
```

custom grow example, show buffs on nameplates
```
!nFvBVTTnq4FmddibBnn29L19H9bNe7wx4k7Ak3UIIjlAjABIkt6jsf7KHQF77UJY6fBhNUxWWqsuOosDphV749CmOvGFaZeWUrCBaBvalg)feXtJgiulSld8E2lVmGntNglsVIh9L4u96a2vjY7VNNgNh6R1jwjiA7W5ZncBGhS87QnErQm2)U1Ia24B2POR1j6uahVw4s8AtpFg985OqBalvVHTMhjWxz8iRuRmKXA5PwyaSK5sL0S0ngg6KcJTPYflePgha(LcWHXIzzZN7SN30DWOEtgaZtVYZs5Tbawlss6hBC61KntCRqzzWxj3gWMEDhM)uMFNXGYYqqzRt43bkVCPJsf0szJ6oauojdatWta3jtXxjm7S0mvLPzDII0jzRuvBDPYksv8KpaBiWhaolqOrKmFKwI696UE(DhdwFICHQ6vWnr6fJaGHnwSOWbcFsIMtb5mJyAuc3ycMrB7iu4QmiCUZ(CZUVuJ8EX(cbXZkYoQhCBrb3wuWTffCV8IxGlMRKR4wbICuMXQx9AiGdH0mffQptj2msBKuy)hZdXeGBff7IZZFBEO7NeDepjpCEk6vZd)L8W)4RvZoxNMhof(8u6dVHB55Hsf83AUm1C2EAnmwx9T7HagQj9xPQlCdVaC1wbyIQ4hC2lOp)aDlNVtX2Lc1HZFSDjzexp1dgokbv9Rf2Yx6PtNa67muPN)W6dXTqzNg4shj5F)m9VFJSG9KGo66E(h6hRuzeP2ZA85ncqNWSfGh(TNuyJxOOpHqjaMQuaxgGZooDO3QT8KpkJTlPD(31qvpjpS1HF5JL4vtfpa2LoSMGx)1Fip8SA5BX4JnUPq4pmvKM88hXfU3MFlSjUJG(j1r)PT)rsYzh6no)PTpuB1poxpn6y5mKZt(VHZBlHX2)(UQWNMh2(H1)r2vFUsxf7WsF4x)V3qp8Wb8akGdHwh7skerTeTsjjFWv7g1nEbql2kpCKCRibwmbPJ7m(oGitgTivNTE36ziVGhsHQtXs8MianPAXUQ9io1yWzXaHlkCQJ)1iI0kG3TGowPvcKPyfxQ(2xDzpbFBRhzZCuTTrRmAPoDpE1mj4WsM8bw8A1pT5(vidTc2ckBpG(ai6yJ7Ct)jms(C5chfRZHSRlONd2fMDanIaHUaVxcwHo62Dm6V6Y2xcUFeihxyE4cInKNzbdA4ABf3n2nfDcmW7f7S4EOUDHf21J721JO2hixbDNmBNT0xrMslufOUVEcZF47Ov2XTVfe3bg(5neWM413V34oVRlThJL1mNuESmZaEpCdKqi6r87rLnFTibGBwZEcpQNoc6KmwVrX2iHTYvvcO8qqdsaDu(P6VZSuV56ei4AOv2OXUt1k3EDP9a9Y1S3WQw5k6eefSg5HpEFMhPHpMRfeFNW3PJHmXN06sxhThPXVVieR7GTQzhJj3Ox5)ZNTIfgUfxlf)WmHt3CyRIwURAxTial2ANAwYHSHFTwAub6zZW5lwLB43BkEn6XXJw3Cng(7LkVpp89z84uI1XVj2FdxBP2Q)uDlLex11ENj(dRHSlpz4e)b996wiVXzK(EEDhp9QH(Wz2X9F9B83Dg6aqUvAKZsOZpLQ3vyUvzjU6(qIbq4tEnVHa6uGBPqUyPfpxxDDH)53qata6Lk(9mHk6o8QaTFHtOB)plZA1QHqDr4gvftSug9f4WSXv5cKmqINT9EL7Tg7KA0tuPIJv)C7bvHow1tSumB07)4p)jXnZTp7Ox4AJRkm6H4jRxYjl8ETEfP7D1Uqh7nsdhcjX0P1mRgJnOXZkPHr8klEa3JbRDyokjrdEKcxs9CaYfvCZ9wx2KusQqT5UZ1JFSaigQoWo7Ws))1UtoQQRC9xmZj5waNN3gHk4p)
```

# Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings

# Todo:
- Test more, particularly for SetPoint issue on nameplates with moversizer
- Add choice for group/dynamicgroup icon ?

# Concern
I'm not happy with anchorPerUnit option redoing layout for all auras instead of doing it only for auras anchored to the frame where there is a change.

Stanzilla commented:
I don't quite see the usecase yet, is it just to replace nameplate addons that add buffs/debuffs?

emptyrivers commented:
It's a feature that i had planned on working on with the dynamic group rewrite, but I got fed up with dynamic groups and lost the motivation for it before actually starting.

mrbuds commented:
It's for anchoring buff/debuff/casts or anything with a "unit" state to nameplates or raid/party unit frames without using custom code and the need to rewrite the positioning logic dynamic groups already has

mrbuds commented:
I added some changes to the cast trigger, detailed in the commit.
It will make it easier to use with nameplate & unit frame anchoring, reduce confusion with bufftrigger2 units, and save Causese from using custom code to filter boss casts

InfusOnWoW commented on WeakAuras/RegionTypes/DynamicGroup.lua:
```diff
@@ -813,7 +814,46 @@ local function modify(parent, region, data)
+        if anchor then
+          regionData.region.anchor = anchor
+          units[unit] = units[unit] or {}
+          tinsert(units[unit], i)
```
If you insert the regionData here instead of the index, then...

InfusOnWoW commented on WeakAuras/RegionTypes/DynamicGroup.lua:
```diff
@@ -813,7 +814,46 @@ local function modify(parent, region, data)
+      for unit, indexes in pairs(units) do
+        local unitActiveRegions = {}
+        local unitNewPositions = {}
+        for _, index in ipairs(indexes) do
```
you don't need to built the unitActiveRegions here, because units[unit] will already be in the right format for growFunc

InfusOnWoW commented on WeakAuras/RegionTypes/DynamicGroup.lua:
```diff
@@ -813,7 +814,46 @@ local function modify(parent, region, data)
+    else
+      self.growFunc(newPositions, self.sortedChildren)
+    end
     for index, regionData in ipairs(self.sortedChildren) do
```
If you extract the whole for loop into a function that takes:
sortedChildren + positions, then you don't need to copy from unitNewPositions to newPositions.

Instead you just call that function per unit.

InfusOnWoW commented:
If others want to play around with it, I've created a small aura that contionously switches from player to pet:
!nF12UnUnq0)LauGSOPbXE3nD7d9HeFjXboYoMkDBrrvmTeLerKjvjPIJ3I6V9odPKS8Tex0xkSHb1O5gN5Wdhh0kWpGOdiDzVeqMhqIWVGiQkCitKysd8(4LxeqEDuCSMzc8G1lBSorXJ8xMZcit6gquYfKCAilWRvazgn85eLSqa(0tkavOHgUuOTH0qvgyHjGeZfCDQBnS0jfwBu8KeMcv3ZMMLcWLrSzfXXUaFBVHJ7)4q492hPfkABia5SSSbrANF1fZyVWegcyf)1aYtDUI4)eX)QjGZkWGsYZOlbNte05SsRSMabGrZGcrTtgRywNqg3B4quVcX6uZ4s(qzwXCX(kgdev1onllESKJHWF0yiZZ4jIasNEE(9MylrwFIQMjP22sHM9uygvRdMz3IHOW5fzgEvvZ92TLQ5FJTTqqmvWNtnmRZcPzLPQufHfcxptXsGEMRshTeko8qCFKdwiHoir46Swhbk2S3sIGwbk8jxNrZcLcOJu2OSwcEyoLloETRrlhN(geuACPgwibDG0ty6dyrjSjNCv3bpsqGNHPe0SFbaCwDB9JvvIk0ERlH(aFoGv8(Sd4)vEeEabEIdyCYcUjmLlsSrjMNucVvaGcWUud1wEPIWuPQpk1vvF0BGF)jxDFpe7AG3nkV6ycITqxFrClFJ4lZwgDhgz5c4a7OV6HqalaRDLBlHtvqiSIWB6mnBODhanCfnIxObBV4clUP(ywsMCr1z)wx4EgnWWE10QVeoeHGjVwT3B(IA1EN0itQC(h)22(7hT)(j8xZMCk7DR0KakukZIKleqfhkGx)2ufWE2dQ1OAVjlWrDyFlALnzocl0gjWHgxiSeDNsZYaWMHP)WQPB85UQvzs4y3QPwTwn9NxnT2KF)KOI5ZxEYFSwzE8QPcPPwDtktS(T7Ofse(et8Y5ywEiTR(SLYqMCIBBDYryJe8TQ0iMzpwWer7Xn7ijxbT8tpHAmu7bjiPLGppBR07dhJV2vY6I8F93h4TNRtLlSQyufSdPuDjAZI2b0omLksyrVRxZvYeMwJCcUsPHpNfDYHuN9Ao3X)5d6zT4gMbxFkG2((vt)8HSSI30AZb1coAlVLhXEJ8ExjBuqoBViKnfD27wepQWU7HgmuLn8D1xXmfkX2BRnaPvQetZ0GoU3bF2y(GWuw4ZalqEeEjA55)QRJWGx9gNOk(H9n8YEMIWoU0lmFNW7LraP7p0YrwB9Kpq1(yzeQZHsQXErj2R1)3nYfWvIS9DS81Z29(J9El5Zmw(v4SiMjiUYomX6jBQ4Vr)(Bn4W3(AroellvoElth4wgyQG5wg5bEKbD758qN3)MeRdkd(1J89hD)Kb3CRF5nt49x9ZOjWCcJE0F4apWXPmEskKvx(Pgkfq6R4FB10hkOrysUAQV)65W(Vp6fUB(1Mdsdp3xX(ZcMiCji582F()3JvzRuhSr5q3ZkmgPy0lmfCxsJji2FX1oQt8PYhstU)o9npSNzLiDM0RNxdpTDVS54Qwifu24HpVSB5UQAqM29e0zzSO6Nxpyt7kyqa57YlhhIlGZ8EFPmWWlC38hX0ud6zw9ashbc9BsGfWoOdWXIgIHGup)nwfgBV71YJpMvbQRs5RX)QqEQ7CZc3mOi2T6SpYl0LRR3FnNgLlGMHUoDBFePRTDMcfrOgOTXmIRXHdgeAPaG)F4p1yOmm9qtU293iM5qlVak(P2O)c(N)

InfusOnWoW commented:
One option would be to remove the border + background settings if the attach to unit/nameplates option is selected. Obviously that's a sledge hammer method, but frankly I'm not that attached to these options and believe them to be very rarely used.

InfusOnWoW commented:
And on the general topic. I think we want that, it's more or less what rivers and I discussed a few weeks ago on how to support the rather common attach to nameplates/units use cases.

mrbuds commented:
Some more example of what can done without custom code
Cast on nameplate
```
!nE1BVjVnq8VlpstAtATc6)2RHcSYen0HdDBpAkGtSj4NMyJCC6)EbF23DNtibkTdvuRX(UZ397UF35g1nkmIveXgiFoILhXe4hylUnzIuN6whfC5nDIyVoD1QcPlkaw)wR1CDYAJ9bJs7Iy9Nggo9EWEB4jYOGlIyX8KNsTMsny3aJwckK4ugDbDToU1blanxP0QI1(1Ws)UWANvLMkTO4bKRwTbUuiJlxTk8TnGvVB4KhgnFcCo9vEPLFb6hYSSXIcVDlkJLpl1ogOL61i2IB7XcxWc7ndmwjEPSnz83aJZ08CzLwKkWfi5zaySZipyLKrypmCYeuUsDJR58oFIjRmxZ8Gr3iMLluLfaU0baUcz2QkylC6dGlNPs1rSBhgeoCgHnKXqmoZWPCszHCrsgVOikMITeCZ8YmNQgU8NE4UfQ3LhUPdtDQCUtsglHN59XyJvGiGpzzLPqYYdXI3aurLGzZnGggi1X0Kujg9kvQhU8gfuQDcMjG8bU5cF6PqcQaPLQSfzfq3CUsF6sVRK50KNYhAWJ0UrqnObIXz9gmEodl4CsRMN9iuOHEEq3FRgiQR07Etel1QeEOy2a46TqncuoYDCc4uqf(T8c32Lg92LybeumbOlOM5Li2JdNfo(2Et8M5VucKzDDn)zeAmVTd6D)WhM0lCiwf7GZMUPMWGvz41yY(rV8Gb)i(jWRSgNhVPkfvougJ2fU0McpepuTTsHCcjzCBg6yDfPgHQDeoGlBFQYF72HYPjpvWWsQGUxCuN0lYOmEArDH0N2YOUE3d33AYmw)vJFUG(9L0VVc098RrZtrSmu(kSoyAWW2XqBd0HmqhYaDABafGhls(YRQlfhnu0DUj)i(y7RGuuil4oSO0tUqe0h))ZE85pHTEmIns91qJhlzWJZNDQ8)peSBLT6j(rzHtkKI7bkx8(eDS5jeN15XO(GPbWU0cN1pt9(7CRWV1idInJSQ33U8pl5cKmSDzy4UsGp)CetWlMC6pxoE2M1E(fv53P87BI7NgV(Ql9EWjeU(kR(G)iSMnnHaCbgdaGWwjayyYXkQ5BeePTS6V4D0uR)vtJuozEaWIPmmMU4XqpY47ncjIGhFUrvlDGImwGs9fJ0ae1bdpA0aLVuxnEcyHMDZQW2q7ndd1HgW1VPzze76dhqIIvykTjY5vc)1JmbX3f1OWnQYGg)Psx9y1MUHhmO8iN78pr4zzOxkcadoRBh)tdosN6JpYPvZ)pBarRroNM0nJCoj5D(otyz3y9ilwKh7lPi4YPsE6Tb1PJ4dh5aPN3nMCQEIyjerKOlT7SsmPtGgqY1ptQfeIY6nyqJ7Ds9D9UxDK0F)3eUFtf(Ry8SwQsxJdpVUQaT)U3xSlAhksL1HYl1tz6uFAf17gFpBuFQ3DHcJGzJ)97QBLmR9uqQ10hWO7Q8Ml7Sxl5)MuXlHsiKaXiy4Jvpbd6HDcqlLuA6siuf47iX2jZORK9tB(03hSDj2U9GOOkdnpC6hFNXhEWa72zdhg0KE8VxPBLjdR7CpgPoRGxe8VOxO0P952)D(4ZA(2zmQPmqYG5mf0ORJ8yDMWSW3VefOAfBJfOM)830Yx2UeMD8TF1hxlK6Nph7Cjphz6qSB)WbnTn(LT)X2Lv)SB1E)hbT))eGVzSkO5uv7S7MoB83NgeIVX6Wh80kLmrUcZi69E2kzTNby7QlWmA0)b
```
Buffs on nameplates:
```
!nxvttoQnq0Fn5ysTM1zsU6pWzCTEahl8UBYbnwaYyvdirqIXFCW)2t3sag7D8UtULQMIb30sQFV(1prhqJOenLmL)kLuqjP4FqOSk1Ek5XWvZ)7WGOrlOewTzNQkS0iusnKIHsogUDRMBObFaxGin6yjNswnfswMajtjswbVmNzGWXSKxGDTwchrGscryjT7frByvg3MkKIM32cVQ3zFh(LPsKLXRW0dSvDta81uEC92TUJ)r)flNTgkyJ9NS6kMhCaL8885PnfUUoM)kxAiWQehOKNNmIe9mjA0kyZQXcGav9ryZ7sDzf3MkzP)cyZTXGdJZYn7C4u3wP1YlLMXfkrLxxijLSeow9ain8kjl)ZaGaoaI9BiPjkyiL0NrAGH0YyPayWeE2fuZtusatWbuWeY3F2Dm77lFyfSCrMKsM4he5VQT)s2P2Vgyl9xeMDbQzvalqJT1Uf(OUixXScQAn)5KCMwBtOKNGblQZnIwAZ91BJQfN43g00YvUtlHL7y1yvvk2ZCYRkEgagNOi9i0HejO(RewHcO2gqRD9eqIuXsf1AAWVdVQ2FPx5WQfClzvyBFm2qLakLMzGigf6RgnD(AcUsJRfIy3vpTdjdEa2RQKfCzgizc(4dqg1cyAiC40pT8tQxoumbugqGXGSuFEJsEEtVbOCrbimdgcTpSwa9pZWObpCvf6GBWON8xUyuKVBU8lIu7jAl7TIm3qWHEtVxgoNlBJ521Lkbk0JcxAxDQO38p0txylkSlWZ32K74WOOWNST0ldR8dgVMV3kIWydMPGXqShhmWZoNOQClc)ZZ(8J2NdXNMBap0EYYbNQ4RnJUQ070SFRbwIsLNQ2lj7faTnUPKMaLemovyNWNhqMpf4XKATrvebFFDzQTFu38)2nXpnZQh)VzRbCiIGjwGh7GtRezWhUJtXlCE5iCiYScvB3W)D873XXudJUtYHjaTf2xzv(9mhVX37oUJx72EXCSXBTNO(TCUFdlulP(kpYf8jvkOx(zGEGVS3jU9G5cbiqr4OnIKxooTXoRLvN8J1w22)vQ4vZ)JhBLVOuDwoldUUmCD0I5bGSyhxKTZ4oE7Q9LS4CEkEOTwF3Xy7T8arxsjWOwPWDS(qS81(38c)Ewf)FQ5YKJqKFX7x))9DjwY8nhYqO40cX1gJsgcmbCpCpRckzwL405n)znlfTaoVjkQXgn(0Pd5d9(6NF6iEFv5oN)q3sVT11)(bRYbeJAMblCERM5V6XZDAaNDf2XpCtByHahOWBqSNkL8t6gqTduKW30DLK39rZjLQWzJvBuOEg3bs3TC37gIZBWb4o69Dj296jx)MBritw57h0KiGLYRV)qiHUJ2svn2M)WJS1QeTrNk0Dh9TxS0Mxl4h7Uyp2PFEf2THE4gs)3)
```
Buffs on unitframes
```
!nBvxZPUnq0Fn9X2jWnn52h5dtINHyOitVTVOGSTWOj2sUsYHqEi)27UswetUKK79TotgIz169JZE2JGoGMsjgkzk)rkPMskW)atmD(CUS0UJM8LRUa(ERDNsVOXkusd4GLsoSy7wd3stWJL5WXlvcjCq6ILZJMbbrR2tAy5CAYakjJL)qPw1kHuKOKC4LYdrJySmT1h2TcPWSZ)m8O3k8Svlkl5A09exv3zaFSGN1UDB6HgiQ3gnF5S1ZHZDFL1QzdHe0WRQIl6kDtBg)rU0sG3s8eLC)KrK07jPJwbbRftkPPIDac(rxxQ5oxjlJMdb3zdsgNvbyerYQ5MqL2kFT0SEt5QQ2A5RGbGtCTKv9xqdbyay7Aeef1mes6JiKcObqJ377hdpxjH(OR9KiscjOMjK)4EFeJ)X8hEdwLOusjtIssJwfM3KDQ9Rb0Y8nHDxIAMgqbAMR2DTpYmQumhHQ1WVpVIzmohA45OX62kRiaB(tFRvJ4z(BnAdyLpB5SQokMsxGZmp9sZlHgZtkkoatiroY)AG3qbqRV5GoUA7B5TqToxudSGSqN6NWaNWYdMCD7sMg5bJXjSeABPDgWQr)xnAA8AsFAFSmST4RZW6ZGROKN6TlbrAROSBbawwgdKuZlBuYx2GuZTAhxJuQffFtu42qHjkAfwjywgn5QtQrpcSojoD2Qr3fbJexVLCj0Oy8ZsF6ITFLV464BWOQ2tj3SkEkaGScrRHM8vFY8Xz1uxfwi6PeOvwpVflFJNKp0nsFDzL)KDyhohirOTbZuWAioJtgm8SYmoVIKSSkEboqkRGkm7dKFcH)umvvvO2lj7fqt4gxvkTV6W)g6(8lUpVe)el(wJvvNcPFDtHBW329)q0IkkDmqbaiyq)ijkdSRmPcOCgNNNOn9rQrVrO5DKJovE7v1OoXSoE)5KjpJELty(rEQ34DQcy48RdUW5QHJ4)eh6L9o6yOh)9jtN)hRR9tDhe2z)tVo7KTUbiaBywmZ8atDYNZZCe8oU74fPPlUBv8n3g2zW9JzvSsO2xSoDECcSaVJlk3bvWqyf(boVzeQNAxHnCiVZfitdxD7jWf2mqhMP5)BlxMFaAKFB4V3lBuYmT45x28NTScSZEztAOyMahddKAhLmoHepnYhnpNkR1AvYfpY1WLN92VpFe7RZ5IoL8lnht0NcB9L3DlG(8D2sCVxNeXRJTPx05OPEkmHRS(jULcVDtclMULI35klx5b9O5vbeuCzQWCmZDDWN27pRu1(vRwRc1tWza549GV7TgVSbLC6gz7e5paKeJ7vyvn78C4V7MdYKvrrj9gNVLo6Uf56rS)yqr94l0pG7mauy6(TaqAomTBD8i9S73wo4IV)2Ka4e6RX(B1Z8O4JWlD5qefO)3
```

emptyrivers commented on WeakAurasOptions/RegionOptions/DynamicGroup.lua:
```diff
@@ -364,15 +364,17 @@ local function createOptions(id, data)
     borderSpace = {
```
Since you touched these options and therefor own them forever, can you implement this TODO for bonus points?

mrbuds replied:
I have implemented same options as group region with settings migration.
Also changed how borders are draw to use position of children instead of the borders of the group frame, without this change for example with grow=left, icons start from the middle and cross the border

emptyrivers commented on WeakAuras/RegionTypes/DynamicGroup.lua:
```diff
@@ -810,106 +818,145 @@ local function modify(parent, region, data)
-                                end
-                            ]]
+    local anchorPerUnit = data.anchorFrameType == "NAMEPLATE" or data.anchorFrameType == "UNITFRAME"
+    local function positionSortedChildren(pos, children)
```
DoPositionChildren is called a lot. I don’t think it is a good idea to recreate this inner function every time.

emptyrivers commented on WeakAuras/RegionTypes/DynamicGroup.lua:
```diff
@@ -809,15 +824,11 @@ local function modify(parent, region, data)
-    local newPositions = {}
-    self.growFunc(newPositions, self.sortedChildren)
-    for index, regionData in ipairs(self.sortedChildren) do
+  function region:positionSortedChildren(pos, children)
```
nit:
```suggestion
  function region:PositionSortedChildren(pos, children)
```
Also I somewhat expected this to just be extracted as a local helper function, and not to a method. But it's not something i particularly care about.

mrbuds replied:
i did it like that to keep access to self

mrbuds commented:
Demo of improved border/background options https://www.twitch.tv/videos/446799805

Stanzilla commented:
Looks awesome!

InfusOnWoW commented on WeakAuras/RegionTypes/DynamicGroup.lua:
```diff
@@ -813,7 +814,46 @@ local function modify(parent, region, data)
+        end
+        if anchor then
+          regionData.region.anchor = anchor
+          units[unit] = units[unit] or {}
```
I think it's better to have the frames per anchor instead of per unit. That makes a difference if GetUnitFrame returns the same anchor for different units.

mrbuds replied:
changed to indesx by frame instead of unit

emptyrivers commented on WeakAuras/WeakAuras.lua:
```diff
@@ -2961,6 +2961,27 @@ function WeakAuras.Modernize(data)
     end
   end
 
+  -- Version 17 & 18 were introduced in June 2019 in BFA
```
if these are in the same submission, then they can be in the same internalVersion, by the way.

mrbuds replied:
yup i'm going to change it

mrbuds commented:
fixed anchor list with non-dynamic groups

added custom anchoring
example :
```Lua
function(frames, activeRegions)
    for i, regionData in ipairs(activeRegions) do
        local unit = regionData.region.state and regionData.region.state.unit
        if unit then
            local frame = C_NamePlate.GetNamePlateForUnit(unit)
            if frame then
                frames[frame] = frames[frame] or {}
                tinsert(frames[frame], regionData)
            end
        end
    end    
end
```

mrbuds commented:
i made libgetframe usable with WeakAuras.GetUnitFrame, this allow to use it for custom anchors
for example with a cast trigger with unit=nameplate + auto clone i can show casts on unitframes with this function
```Lua
function(frames, activeRegions)
    for i, regionData in ipairs(activeRegions) do
        local unit = regionData.region.state and regionData.region.state.destUnit
        if unit and UnitIsFriend("player", unit) then
            local frame = WeakAuras.GetUnitFrame(unit)
            if frame then
                frames[frame] = frames[frame] or {}
                tinsert(frames[frame], regionData)
            end
        end
    end    
end
```
```
!nB17VTTnq0)xkWasb6cQDBZ6x2h8ptDHRCQP8A3gMIPKOT5ImPkjvsCgM)BF3DuuXrjPjaBFyio2sKhpE3JV7DmPtsCcZMWgkUmHTnHLJFGH4MSPc1A3MKO3CYRtyPAtUW0NNDrUrxMW6xiV5gUjF)YyTUWjHHUE2QvwHljcmF3bpV2iZJ3vksyZhgC0aDH2a7tuh6tx673qF)wyrh)UehewL8mbojJN5KALLIvh34GhG5Lkz9tRGhTBONH3Cg561cJ17)4MbWhZfPvRw5dNpmA6zJxmfMNELxz4DXDvuumj369STkvCPq5yWQKxNWoFqpw85S4EZbNvHbaRSGVdCEJPNzeKPSZgnfCongSzcEbGMmfFRWgI0k1THMZpuMUOARI5t9oys6egfV4xGecWaySFccvLClhHKdrK60qPvW35qYGgCUFqRitRGCc2GTCP65BDdY(8ShwbVqUwLWgmkkE0Ck6OeezcfAorUQSIZZk4wBskH3z4GBRaAuay8Z2EuR8gr7bHHtRzLplsLh6e0oNXl8GCwL1P32tLTbCaBvLI4BhTYGhwVA)sK)DPyUynYcF5(pUFP)VvAZ(LsWadn1qUJdVRG)l5sJ9OwRBzU(21g(RqdHX(LixA)YF(qxDS)XJbi0jGGqL)OZECUW6wGU4E(xUk4CYbOrtSJnsHk)Ox45UV4vEtGi0TrOUVpAfReUqb7xe8l6bKb7XNkO9Fmo1rEN9OUbdPAF893VgCMoi(D6N)G24wJGNd)1F)0oYjvwHXD0Dw(Do)(oHnGy3FY7mi(syMpcVeK7q2M)Pr5RHcMiQkYVPETO8DGWGmBTrxvgSLH09OoNamvnuHxx7z45YklqVFnPTQVc0wMnFYVnlkU3uu)qbmcLBmW9qY88EdNSGH8EKCFMwIIrHIt)(e0QFhQb6aRMvgKBDE57ViZXobGbsiNgEkYLq3bm6aVBPt7jrEOn47jkY1DGImAcqlga5KOtcrery8yWGfS4zFcLtkw1kqHSALCTpGQWyy7xF5jxm700pi3b6kYTsF8dWXTsNOOK8GmbKDMswMIg68cOGUeQ3CBVcX1UUT2DCSoJ1qxa)jc0LyDH(QqdYoWbXfcrzpukZnh9lUfOj4Vh2l8bpf4fLB41rSUixFLIDLeaK(uiJUXRSLgeQIH4zrzokJXQQ)nSuIGL(FzlZAnEaVaAxT2nYe7)eDhdnjsyDA6coGBDpANs01ul1(h2MSvdz0iEkuqK(jDUhJOUW4ca1mcsatc8sekQu17oWU1p8vaA1l(E(03zJuZJ9wr7E0p2HOpsGQHHYd2Sgpa)6D4a)pU5D7As2G5Jgfvx0fF932m)MY38(p)N1ffJu80croM707dGIeie3sW9Ki2KHJ8aW3V3ChIdIoOU0O)Syqky(Kt)qOKel)gxWxduWzlINojcC8gHC9ghkf(GIgunQub8OO3txS6PIaCbJnIVvjuz7WBm09D39ATuK8O5OhXsRCoTA2Lcd0BTbwEEzpMKjmO58n7x(5kEUHUqrmXtKzxSByOKkTjwEshFyxgINsX6VEqwDhH5ohG3E5UUV92BU9V)YA3O1BBWsyZ(bBn4Tbsq4SYsHau(cYnM6lQrbeyAzDjCF)syn3FdjNnfR7xMrQmniABAd5UaZfC)v(UBilkOIIkSdL2gtoS)JubNTwVi79Bx2aEp85y7EsHnmKt9d3xGaRlHS7TDXJXK)5
```

InfusOnWoW commented on WeakAuras/RegionTypes/DynamicGroup.lua:
```diff
@@ -931,34 +946,34 @@ local function modify(parent, region, data)
+    if data and self.dynamicAnchor then
+      local frames = {}
+      if data.anchorFrameType == "CUSTOM" then
+        region.customAnchorFunc(frames, self.sortedChildren)
```
I think a better interface would be...

InfusOnWoW commented on WeakAuras/RegionTypes/DynamicGroup.lua:
```diff
@@ -931,34 +946,34 @@ local function modify(parent, region, data)
+        for i, regionData in ipairs(self.sortedChildren) do
+          local frame
+          local unit = regionData.region.state and regionData.region.state.unit
+          if unit then
```
to call the customAnchorFunc here with either the regionData, or the state. I somewhat prefer the later as a parameter.

A third option would be that if a region.state has a "anchor" member, than that's the aura's anchor

InfusOnWoW commented on WeakAuras/WeakAuras.lua:
```diff
@@ -5678,14 +5679,25 @@ local function GetAnchorFrame(id, anchorFrameType, parent, anchorFrameFrame)
+    WeakAuras.ActivateAuraEnvironment(region.id, region.cloneId, region.state)
+    local ok, frame = xpcall(region.customAnchorFunc, geterrorhandler())
+    WeakAuras.ActivateAuraEnvironment()
+    WeakAuras.StartProfileSystem("custom anchor")
```
Stop

Stanzilla commented:
Thoughts so far:
- The Anchor per Unit option might need a better name and definitely needs a tooltip with a description
- Icon for groups I like but the option should look the same as the single aura icon selector
- What exactly is the default icon?
- https://www.youtube.com/watch?v=QqN6yIgtlq4&feature=youtu.be at 1:00 you can see one icon fly in from the top to the vertical buff tracking group next to the main cluster in the middle, that is flying in from way too far away.

Stanzilla commented:
![image](https://user-images.githubusercontent.com/75278/61461105-550f2d00-a970-11e9-93aa-10e8124aacf7.png)
small regression, the combustion icon is not in the middle anymore 

```
!T35wZjoUsa4FnN6mZd7wyUZdZPQaHCz2eitmzMSBTfbJraEIXMyBsc5b(TFAjzdY2YGTXMBXvT1oeFrwxA19NAj1QRq3oDfLG)NvxrZUINJETR4KUIdW)h(ogY3G0gznUBRcLZ1vSVUXaKrDj5NhyOpTRyDvLp(qYyWIED01vTuGljnZASUr7Pwk6AM0eEKH(SPxlRR1vmFHYflwbEknz4PUtxrdUFJMT608ENuVHUQUb8ITe62kx3w5j))cK)Fr8fX50PsYi8TeLKD(oIMwsgw0pOIMI9Vgc)0Cm53WFzzOmAeYWKM6DwEbYpNpfHZ9gsqsnRp6vKML4SHdvEVR4tnotSZtIDo7E4bhapc(J2vuORi5X6kEfssfQKw(I3zGiVO4DnV5gC(fPQE9a7QJbO(q62H89UQ5n3DXdWJOjnbzF)z4CV4uvP5qod(l7VHLYee0UGFxXX6V1gYaMK)9mOo4ve9nxvGifyjYT6qV4T6dq2vGY6QZMOjsRgfWvywidnj1Fcvo4IwR8c42iLj0sktT7YI)t0kmte0UcLm76pnDnsozIKIw4F6LTsH75XfmvLrARKCGChPydcjIQ6sezxfnICh(NtMbYMlfcKuX1NEVSSQKjvocErv4JD7zx2K7RM337oqbKtKHRm33Tm0H0Y7fhsfB9D9PVonGChibj7pTXnFEV4mt0t0YciDpdxhQ8HVNcEU(2DIz7Ujq6Ujq6Ujq6UL7plrQUjIcOU9HKtgYHeHgA31R1mrwKkE6FJFg6VAoye85fw07oL3rQq2fnck0u5(bZbbEfzAlK9ZlIZO4o1QktGEaTkH1COmG(c3Fo8P1brqQqG482dhs(W)HqTCybAnqgqZ6cOMfkoI3F25x)GiP24gsQTmx58EcLXnoVTQpGLU8RoI)vZLphKQkq)TBLgHUv636GwpdSaP6qpATgAaDDH6Ajljs6qvTDb(Q0SUyJ7B2SfTW8lLbyLPqrBgoXZD2OVRjw(Yl(DzszyOYOLQmFd0o0((R)N2T6CgOG4DNmoT0oqHrbRHUfTJk(MgsduMzcvKqjalxVuz3ivin7t)xhT6c5yQkZfGEzzq3(a930eFtbkp1HlmZ0sFsh07wpmDawUqCM9)68OK2E4BPG16xF9AEDPqDesdzOiZQH7m13KMBUELRTGABiLZNZrvbinynZCdQYxQBfmAPt76qvPIZYCvtB)8nSlMGSTH(idKP5IEFreNr(knzi6WRJL5LufxMbH3tdTON(Wf9GUDdS)IlVDDVwkc2Oa8EURPQtViigk)8QMONgRa68TnD4XaHT1d31XRR63D98ATVW1EYZi00ZWkYSUh3EJfp4BIHrzFqgeymXeUNELjMq98eZ0MGiKHTApCNM1RQuGMPmFzgQPMuFvOghAhgJugng6BvO0klt4gQPiR(swwQKepg2vcWoboPFfnwrMMWoMnSTPbz2qybjyRqRSTupalB2c0Y4hiiRCm5XhuikjcWsDymjhG1nokZXTGxyGGMhn55ylB5lrViTlE)zww6ATH3c6JzFdip(Sg05g3(Yt1p(HEKr7jOS5EIrog9CGssqPI0a3pPT82S(47B)u0FUON9FkVsA7IkffQuUwHs5kulxXcaa9W)OurkB8Fwf3SEr1IvQaS15YvSy58c5iprHvYLKuCOooZF7CdfjGy)lqMhzilzI(QRSzi4VzE6)MTqrU8QQPZEOtB7R(QIPc0LGyka)3USZ0(bO28P6T70P9ToMGytxI5nuhs1tR2TAYuCOmdyB50x7rVzh8drBGHVYnxdVmvWNHfHyJIAvubATbbg32I5nIgv3LCj1PJPw)nTazM5N7ysR)A4bCSvIlyNRysvzaVWh66tOPk2QuD6xtCj2fMBOHEJgNtW4dwH0fvYNRuLs1kjiiuUCTkLPIew(jwEJcLGvrXszqiuq)69xE(7vNj)8l4oQ4ovenl8h4fZWTwnimAx(vkq77bGH2rrEmcSCXqgGZzVkbYQy5grDnhBU4RQoB5izglPncTkzO3J2RPATkcvH(n5kwjF(I5QTSxtq3PqG3XrYFkqmJma9X219wokCCAmDAVQ7adBH)Gm0xCKMoYbW4ckLuuyC4y2wiSZnKgPR9FHNUoCh8W23agMB2lwOSnIHXanva6fkKGqyoPxQHHDymsF)kutsuSGqpWnIyXsQ2QqWH4gnkObM7GgPcjnmI4c55dWeiqionOVB94IaMFl9DrW0HKm2ty9JsKXG))(ghu3WcfgbgowBwz8Cz8CBINtEZ6o8Y35LyY9yddKWBjjxqOEbt4rnsz6yHYRpBwbCYaVra2ot8fXQ)L0xhC(DHaylipL)5gyRalW2PJ7XYddhOOGx2mp4xRhvl19q2DJ1rAkVVO3fQeoRuKnlDDrwnqXD5tFVKDeHNzJTupMWAcyaKKGwZHeti)NDYTa8SYHfa3)XmdGBFaW5MXkIaCmurRqLE3NL8KLzZ22f1Yf2)CEP28Ir6fZ0nZkbNBwH)UZG3L(1VKQN5)TGW5OEyBdLaiNAGZP2zrmnfEjSa66wFXiirSyrwwCp6kWOjYKeiPSCEHWlHXNen2Ej0f9Ohu0BrwiDJOmjTbGGhj3ekuQqjo(jmKuNoVCcqDYX8A6nLO7eqVnnHLrcUZpphUE05rGAI64QG6jnzNVjpD9ZwBOi44raN((F7qEunzocmb5i70(U4brwnn9cOx2mUeJry2C3cUrQHM7rsdMVzgXqa)sah)yY3)DJZRFvZ6x7cC0XLtFU9WxjwomwfyhlE7JfTY9kZQKpp8Xf9ATZ6kdtZwJr18vDz7I0giPcYdMBId8eXnFNEax7lpRXsIfg(lHT0ZA(GY2khQfir3oZZAF2qZYCX3r4C0Uhb2ymPfvx8ryZEEYpE5hxDU1G(pN5uVJFN6v(uy3v4MHJNd18SK(C51PY5lhjh2XfBlmOMCzz9qCEgOHfVplAB0pj8ENFq5LSLmu4httD8HXk7ldU9qdUncoxCDZgSlkxEOW8GOpQ30gzaKhbBAdgGrVuA7gSrQDjIvP35U5rIRV(0N22A0DAZ1FSr2A(lz4jTPixXvUd5jRSV5j9kjWL2WNEu)QAt(fhwXQE70hwJg(PncPjKa31M7M9iqXegFO4oyskxh)1g9tw0zocaNGhoCKui63EZHmoDeDFsGg4wzoIJnuw7wBHLWg6t6dALG01EwV20cU322DrNLf193)1o)qEyLhVSQllEP5uRulxbHAha7ST1nzlRQy3jRBfEJgpjhimPg3wBJ5VNPrJ0nGS78GhAC2C1Fumx9FUMV5QNcUiKrJry2ZVUMv6I53G6Wy7rqE6rdWHG3H7Sf3PH2LzHn5PZvEiKvB9(w14bZmrJTJNXvNst(C0JUy8WWdGHpDCtx2AbmJVypUwad2zHyfv7tNfABXkzDx4CRFo5h)8Rnm(5nzt)Cmb6QXc0fZ4WZPZolgumju8Wo07DHH(hiTipPWhuXDpA18Ud6B)S5EJoY3M2yh2iFLQS7d3kfcpDx0CgQpSzUGVPWM2nJ)5yz(spraW9U)Iyl(rIYBZE829OCzMW4TaKJA4bm7G3rWCind1McMXBr0z8n)B6skLLLhEvgcxmr4GkZtlNYfIfTxSrYC5npOriC8ybqo6bn7A55l69tKIwic3lmaErfnlLNScsTsATrqsNzspLJcEP0(TnM4xeVYLMBV2OaGfbERqgUY2kd9hgZGDMN2(uSRBtGaR3we6KbJnetn2l8a(EwJX7FCi2ieyI3nQ0l3v88R(zUS1CxCjWecTx0OFLJqOmweHCvRcAroSxbjxOyGw0Roye3ks0yhQRHeADohFMPQMfz(o6cCY7aKXTjySeiY7oBJ)MT3iY2BeBccZRd1cpqwY4rSR1gcAx1TTXeCOnoyh2fat2xEUvJlF5dl9bfZyYo(3xTc5pj2yTh6bbVW5KUBvmm0nw076js4J1V0BguzG3kvQqX8Xmu458UzXF5JXWY3wgk(QMaBYLu3FHPu44l7iWiZhCPp(3w4boQLeIDeBNWfMTRu8Pc)v)5LBB84pQ8W1zZvACrXkSVqXsUt47TZ5D5Zvkhw9AicTE7I4Gsi21dapNjyyW(aN9wfTbr800W9KTgCuqHpa5bk22PASpjE7k6mZ7hfM3JctNVbeWHYZz8T(XgcHP4Gyjc0WSxOI0C0f93rEuM3inQIhPrC37pXB4gCh8uOJnpRZHZBZ5GH(TbeE(8DU5gaY27)LzrPHvA(7hl5czZ7Ot(KdLv8ypkghTDu6E48Hi49tWLQsYksQl6jov5zuS3uPX35yc1QvPA5mVJ9j)0jljp0kIUbSchc(l7ulAYLbuVh8x2wDCJfVGPtYmBQ2gISndLah5yTMC3hV97F9qZZEiZFzB1SOsm4(Gj9p2CrOo7hJG1fomWsF29nxXIfX7TVJhxZDJYRGb0f9QRpPFsepswBmkEDrGKWrlsRFZoKl2hNQycv3hNQyrEZKgA)kKTThY22dNOiK(hT5gcYiBA2xtJ4kc10d2WdE)OU2T)G)PvGoFQwZ(hRrpEP0hIzNSy(PXkFaTNgsoanSoEqdk0Wn4EG2GSy2I1jlb7uQkuSCbFHnehepSibt4S0kTIFCHB5V1SVUwFD1nVNi4IDYdjmcXqUGpLjch)MDf9j)SQEIUy4ssAs2f4gOQk6E5Rqi8Yx0JaDHBZ0Mf1sYMO715m2aqSIBadznhyeHCFQUnNXy2gCcMslubkKQLFP(lLQQPGMLbR5hwRY(CgupCpMfI)ckARpMfoMxuiRSSI9ttHOBDnvdpfryPIW)0tyNm(GnheB2GJp3zdjaK9rFGI3acyGZluOqTsrz4aboaP4n4a63FTJniwltmI5NR((Jc1fQA0(lxg1rJg6LtMx7EzOEhfOEBZcmdz1s)vjmAKVvugpNyXHlKLZJDoopSpGx(mHMv9il2GSfwad2LxP6QIlaZCB5cJdkaiRf9aIrJjrCdh4(mLTe2SyIeXDtmdHhDXs3JPLIg7mDM2CZrzTKLoZVzgLYPMdPoMoel2KBWiW7x)8yHc3j(yb9ACzOcnHEAeeriMzigzIVhYcccmBHUfEmTA7dpOLCZRPnocqOjuoNGGlMTCCjhARDMDieBJNlcIyqq62V0(G8GN3nywtB4FYeb4)23cGaCjZNDsavnT1uNtuzsEs7BVeLBcy8ArVB0hXleUPzJM5JvtJFHG8C6ZmKrpS(ZdRhA0OPOOpkVvV6YRnCMQQPSKM)P4C1ITKTvJFnpTSlbTG5P5sALxdcbtF6DyzT9wYSCJIZ52uaFt(Yd212xb8WGggpLA(TJyAlq22szQnHdttdtJk8CoTjYqBjOM(wAgHMfUbYEMwBG9h)55MV5uHZdao)b3Y8JdQ1juGlzDiW7e)fhTWzmxi3qmHCraWYLLvRd)f1hlQbpxOUc(W)Gd5U9jZgZWrZygyNr5WePu6ZJXfBaQoO2ElwPGEPtPRGV2ZbgWXxMFEnxGVHWlBoEcBGIj2Y21er5RXwfavdO)f)3M)lXo2t4a8YtgqrqF4uB6NGbjHK(v8oAnp(R09)d
```

mrbuds commented:
- Fixed positioning issue, couldn't reproduce the flying animation from your video with it, but i didn't try before last commit
- Default icon toggle removed (default icon show when icon input field is empty)
- MoverSizer position is a bit off for dynamic group with 1 tsu/auto-clone aura, i'm not sure if it's new or if it was already like that https://i.imgur.com/RR52oWW.png
- Do you have a better idea than "Anchor per Unit " ?

emptyrivers commented on WeakAuras/RegionTypes/DynamicGroup.lua:
```diff
@@ -277,17 +326,27 @@ local growers = {
+            x = x - (regionData.data.width or regionData.region.width) - space
+            y = y - stagger
+          else
+            newPositions[frame][regionData] = { [3] = false }
```
This does too much work. If you've exceeded the limit, then don't add position data for that regionData object. The positioning system is smart enough to detect that you just want to hide it.

Setting the show value to false is useful only for making animated dynamic groups behave slightly more consistently.

emptyrivers commented on WeakAuras/Profiling.lua:
```diff
@@ -6,6 +6,8 @@ local profileData = {}
 profileData.systems = {}
 profileData.auras = {}
 
+WeakAuras.profileData = profileData
```
This is unrelated to the rest of the pull request. Did you want to split it out and submit it separately?

mrbuds replied:
sure i can, there are also a few other UI related stuff that can be splited, but the border changes are a bit too intricate with anchor changes to be splited

emptyrivers commented on WeakAuras/Transmission.lua:
```diff
@@ -1642,11 +1642,18 @@ function WeakAuras.ShowDisplayTooltip(data, children, matchInfo, icon, icons, im
   end
+  ok = pcall(regionOptions[regionType].modifyThumbnail, thumbnailAnchor, thumbnail, data)
+  if not ok then
+    error("Error modifying thumbnail", 2)
```
This isn't actually all that useful. It would probably be helpful to pass along the error message, so that if we ever break the thumbnail functions, it isn't a nightmare to debug.

Or, alternatively since this halts execution, don't use pcall at all.

mrbuds replied:
Should i remove pcall for createThumbnail too ?

mrbuds replied:
Thinking more about it, because of Classic, and high probability that sometime someone import an aura using an icon that is not in game files, I'd prefer if it fails silently and doesn't halt the display of import tooltip

emptyrivers commented on WeakAurasOptions/RegionOptions/AuraBar.lua:
```diff
@@ -85,15 +85,15 @@ local function createOptions(id, data)
       order = 38
     },
-    symbol_header = {
+    icon_header = {
```
This also looks to be unrelated to the rest of the pull request. If/when you squash, please don't accidentally squash this into a commit which says it's talking about dynamic group improvements (or, just also mention in the commit message that you changed this)

mrbuds replied:
i tried cherry-picking in another branch, but i'm getting conflicts with master, so i'll make no squash it

emptyrivers commented on WeakAurasOptions/RegionOptions/DynamicGroup.lua:
```diff
@@ -75,6 +75,28 @@ local function createOptions(id, data)
   local options = {
     __title = L["Dynamic Group Settings"],
     __order = 1,
+    groupIcon = {
```
Add a tooltip, please.

mrbuds replied:
used `desc = L["Set Thumbnail Icon"],` for dynamicgroup & group groupIcon if that's ok
